atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java

```diff
@@ -463,6 +474,189 @@ public String toString() {
     }
 
     @Override
+    public Map<byte[], RowColumnRangeIterator> getRowsColumnRange(TableReference tableRef,
+                                                                  Iterable<byte[]> rows,
+                                                                  ColumnRangeSelection columnRangeSelection,
+                                                                  long timestamp) {
+        Set<Entry<InetSocketAddress, List<byte[]>>> rowsByHost =
+                partitionByHost(rows, Functions.<byte[]>identity()).entrySet();
+        List<Callable<Map<byte[], RowColumnRangeIterator>>> tasks = Lists.newArrayListWithCapacity(rowsByHost.size());
+        for (final Map.Entry<InetSocketAddress, List<byte[]>> hostAndRows : rowsByHost) {
+            tasks.add(() ->
+                    getRowsColumnRangeIteratorForSingleHost(hostAndRows.getKey(), tableRef,
+                            hostAndRows.getValue(), columnRangeSelection, timestamp));
+        }
+        List<Map<byte[], RowColumnRangeIterator>> perHostResults = runAllTasksCancelOnFailure(tasks);
+        Map<byte[], RowColumnRangeIterator> result = Maps.newHashMapWithExpectedSize(Iterables.size(rows));
+        for (Map<byte[], RowColumnRangeIterator> perHostResult : perHostResults) {
+            result.putAll(perHostResult);
+        }
+        return result;
+    }
+
+    private Map<byte[], RowColumnRangeIterator> getRowsColumnRangeIteratorForSingleHost(InetSocketAddress host,
+                                                                                        TableReference tableRef,
+                                                                                        List<byte[]> rows,
+                                                                                        ColumnRangeSelection columnRangeSelection,
+                                                                                        long startTs) {
+        try {
+            RowColumnRangeExtractor.RowColumnRangeResult firstPage =
+                    getRowsColumnRangeForSingleHost(host, tableRef, rows, columnRangeSelection, startTs);
+
+            Map<byte[], LinkedHashMap<Cell, Value>> results = firstPage.results;
+            Map<byte[], Column> rowsToLastCompositeColumns = firstPage.rowsToLastCompositeColumns;
+            Map<byte[], byte[]> incompleteRowsToNextColumns = Maps.newHashMap();
+            for (byte[] row : rowsToLastCompositeColumns.keySet()) {
+                byte[] col = CassandraKeyValueServices.decomposeName(rowsToLastCompositeColumns.get(row)).getLhSide();
+                // If we read a version of the cell before our start timestamp, it will be the most recent version readable to
+                // us and we can continue to the next column. Otherwise we have to continue reading this column.
+                boolean completedCell = results.containsKey(row) && results.get(row).containsKey(Cell.create(row, col));
```

Comment:
Should we just do a single get and check if it is not null rather than containsKey to avoid traversing result twice?